File: frontend/src/chat/voice.ts
/**
 * One place that knows how to hear and speak, whichever side does the work:
 * the server (`/api/voice`, when STT/TTS providers are configured) or the
 * browser's own Web Speech API.
 */

import { create } from 'zustand'
import { record, type ActiveRecorder } from './recorder'
import { guessLang, listen, speak as browserSpeak, speechInputSupported, speechOutputSupported, stopSpeaking as browserStop, type Listener } from './speech'

export interface VoiceCapabilities {
  stt: 'browser' | 'openai' | 'gemini' | 'elevenlabs'
  tts: 'browser' | 'openai' | 'elevenlabs'
}

let capabilities: VoiceCapabilities = { stt: 'browser', tts: 'browser' }

export async function loadVoiceCapabilities(): Promise<VoiceCapabilities> {
  try {
    const response = await fetch('/api/voice')
    if (response.ok) capabilities = (await response.json()) as VoiceCapabilities
  } catch {
    /* keep browser defaults */
  }
  return capabilities
}

export const canHear = () => capabilities.stt !== 'browser' || speechInputSupported()
export const canSpeak = () => capabilities.tts !== 'browser' || speechOutputSupported()

// ---- hearing ---------------------------------------------------------------

export interface Hearing {
  stop: () => void
  /** Resolves with the transcript ('' when nothing was said). */
  text: Promise<string>
}

/** Listen for one utterance. `onInterim` only fires with the browser recognizer. */
export function hear(lang: string, onInterim?: (text: string) => void, onLevel?: (rms: number) => void): Hearing {
  if (capabilities.stt === 'browser') {
    let resolveText!: (t: string) => void
    const text = new Promise<string>((resolve) => {
      resolveText = resolve
    })
    let finalText = ''
    const listener: Listener | null = listen(
      lang,
      (t, final) => {
        onInterim?.(t)
        if (final) finalText = t
      },
      () => resolveText(finalText.trim()),
    )
    if (!listener) resolveText('')
    return { stop: () => listener?.stop(), text }
  }

  let active: ActiveRecorder | null = null
  const text = (async () => {
    active = await record({ onLevel })
    const recording = await active.done
    if (!recording.hadSpeech || recording.durationMs < 400) return ''
    const form = new FormData()
    form.append('file', recording.blob, 'speech.wav')
    form.append('language', lang)
    const response = await fetch('/api/voice/transcribe', { method: 'POST', body: form })
    if (!response.ok) {
      const detail = await response.json().then((b: { detail?: string }) => b.detail).catch(() => null)
      throw new Error(detail ?? `Transcription failed (${response.status})`)
    }
    const body = (await response.json()) as { text: string }
    return body.text.trim()
  })()
  return { stop: () => active?.stop(), text }
}

// ---- speaking --------------------------------------------------------------

let audio: HTMLAudioElement | null = null

/** Read text aloud; resolves when playback ends (or immediately if it cannot start). */
export async function say(text: string, lang = guessLang(text)): Promise<void> {
  stopSaying()
  if (capabilities.tts === 'browser') {
    if (!speechOutputSupported()) return
    return new Promise((resolve) => {
      browserSpeak(text, lang)
      const poll = window.setInterval(() => {
        if (!window.speechSynthesis.speaking) {
          window.clearInterval(poll)
          resolve()
        }
      }, 200)
    })
  }
  const response = await fetch('/api/voice/speak', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ text: text.slice(0, 4000), language: lang }),
  })
  if (!response.ok) return
  const url = URL.createObjectURL(await response.blob())
  return new Promise((resolve) => {
    audio = new Audio(url)
    audio.onended = audio.onerror = () => {
      URL.revokeObjectURL(url)
      resolve()
    }
    void audio.play().catch(() => resolve())
  })
}

export function stopSaying(): void {
  browserStop()
  if (audio) {
    audio.pause()
    audio = null
  }
}

// ---- voice session state (the orb) ----------------------------------------

export type VoicePhase = 'off' | 'listening' | 'thinking' | 'speaking'

interface VoiceState {
  phase: VoicePhase
  level: number
  transcript: string
  error: string | null
  setPhase: (phase: VoicePhase) => void
  setLevel: (level: number) => void
  setTranscript: (text: string) => void
  setError: (error: string | null) => void
}

export const useVoice = create<VoiceState>((set) => ({
  phase: 'off',
  level: 0,
  transcript: '',
  error: null,
  setPhase: (phase) => set({ phase }),
  setLevel: (level) => set({ level }),
  setTranscript: (transcript) => set({ transcript }),
  setError: (error) => set({ error }),
}))

File: backend/app/api/voice.py
"""Speech in and out. Which side does the work depends on STT_PROVIDER / TTS_PROVIDER."""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, File, Form, HTTPException, Request, UploadFile
from fastapi.responses import Response
from pydantic import Field

from app.models import CamelModel
from app.voice.providers import VoiceError, VoiceServices

router = APIRouter(prefix="/api/voice", tags=["voice"])

MAX_AUDIO_BYTES = 10 * 1024 * 1024


def _voice(request: Request) -> VoiceServices:
    return request.app.state.voice


class VoiceCapabilities(CamelModel):
    stt: str  # browser | openai | gemini | elevenlabs
    tts: str  # browser | openai | elevenlabs


class TranscriptOut(CamelModel):
    text: str
    language: str | None = None


class SpeakIn(CamelModel):
    text: str = Field(min_length=1, max_length=4000)
    language: str | None = Field(default=None, max_length=16)


@router.get("", response_model=VoiceCapabilities)
async def capabilities(request: Request) -> VoiceCapabilities:
    """Tells the UI whether to use the browser's speech APIs or the server."""
    voice = _voice(request)
    return VoiceCapabilities(stt=voice.stt_provider, tts=voice.tts_provider)


@router.post("/transcribe", response_model=TranscriptOut)
async def transcribe(
    request: Request,
    file: Annotated[UploadFile, File()],
    language: Annotated[str | None, Form()] = None,
) -> TranscriptOut:
    voice = _voice(request)
    if voice.stt is None:
        raise HTTPException(status_code=503, detail="Server-side transcription is not configured")
    audio = await file.read(MAX_AUDIO_BYTES + 1)
    if len(audio) > MAX_AUDIO_BYTES:
        raise HTTPException(status_code=413, detail="audio is too large (max 10 MB)")
    if not audio:
        raise HTTPException(status_code=422, detail="empty audio")
    try:
        result = await voice.stt.transcribe(audio, file.content_type or "audio/wav", language)
    except VoiceError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    return TranscriptOut(text=result.text, language=result.language)


@router.post("/speak")
async def speak(request: Request, body: SpeakIn) -> Response:
    voice = _voice(request)
    if voice.tts is None:
        raise HTTPException(status_code=503, detail="Server-side speech is not configured")
    try:
        audio = await voice.tts.speak(body.text, body.language)
    except VoiceError as exc:
        raise HTTPException(status_code=502, detail=str(exc)) from exc
    return Response(content=audio, media_type=voice.tts.mime, headers={"Cache-Control": "no-store"})
